pkg/concurrency/jobresult.go

```diff
@@ -0,0 +1,69 @@
+package concurrency
+
+import (
+	"sync"
+)
+
+// This is a Read-ONLY structure that contains the result of an arbitrary asynchronous action
+type JobResult[RequestType any, ResultType any] struct {
+	request *RequestType
+	result  *ResultType
+	err     error
+	mu      *sync.Mutex
+	done    *chan struct{}
+}
+
+// This structure is returned in a pair with a JobResult and serves as the structure that has access to be updated.
+type WritableJobResult[RequestType any, ResultType any] struct {
+	*JobResult[RequestType, ResultType]
+}
+
+// Wait blocks until the result is ready and then returns the result.
+// Returns *ResultType instead of ResultType since its possible we have only an error and nil for ResultType.
+// Is this correct and idiomatic?
+func (jr *JobResult[RequestType, ResultType]) Wait() (*ResultType, error) {
+	if jr.done == nil { // If the channel is blanked out, result is ready.
+		return jr.result, jr.err
+	}
+	<-*jr.done // Wait for the result to be ready
+	jr.mu.Lock()

```

Comment:
mmm not sure why we need the mutex at all? arent' we using the channel to block already the `Wait` call here?

Reply:
I think that this was initially defensive code for the situation where two readers were waiting in separate goroutines - but looking at this with fresh eyes I believe you are right here - the channel block will be sufficient and will simplify this code.

Reply:
@mudler I remembered why that defensive mutex was created - although I'm not totally sure if it's valid.

While the mutex is checked here in `Wait()`, it's also used in `setResult()`. Theoretically, there should only be one reference to the `WritableJobResult`, but it's obviously possible to send that pointer around or duplicate it. I was using the mutex to enforce that even if this occurred, only the first write would "win" - closing the channel before `jr.result` or `jr.err` were overwritten.

That said, rather than using a raw mutex, the `sync` package has some other options as well, so I'll take a look.

Reply:
I ditched the mutex in favor of using sync.Once - while the logic should be unchanged, I think the new form communicates the intention more clearly.